File: src/aihil/debugger.py
# Copyright 2026 Hannes Pauli
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from typing import Any

from .config import AIHILConfig, ConfigError


class DebuggerBackend:
    def info(self) -> dict[str, Any]:
        raise NotImplementedError

    def probe_target(self) -> dict[str, Any]:
        raise NotImplementedError

    def flash_firmware(self, artifact: dict[str, Any]) -> dict[str, Any]:
        raise NotImplementedError

    def reset_target(self, mode: str = "run") -> dict[str, Any]:
        raise NotImplementedError

    def classify_last_error(self) -> dict[str, Any]:
        raise NotImplementedError


def create_debugger_backend(config: AIHILConfig) -> DebuggerBackend:
    if config.debugger.type == "openocd":
        from .debuggers.openocd import OpenOCDBackend

        return OpenOCDBackend(config)
    raise ConfigError(
        "config_invalid",
        "Unsupported debugger.type.",
        field="debugger.type",
        value=config.debugger.type,
        allowed_values=["openocd"],
    )

File: src/aihil/debugger.ts
import type { AIHILConfig, JsonObject } from "./types.js";
import { ConfigError } from "./config.js";
import { OpenOCDBackend } from "./debuggers/openocd.js";
import { STLinkBackend } from "./debuggers/stlink.js";

export interface DebuggerBackend {
  info(): Promise<JsonObject>;
  probeTarget(): Promise<JsonObject>;
  flashFirmware(artifact: JsonObject): Promise<JsonObject>;
  resetTarget(mode?: string): Promise<JsonObject>;
  debugStartSession(artifact: JsonObject, mode?: string, timeoutS?: number): Promise<JsonObject>;
  debugStopSession(timeoutS?: number): Promise<JsonObject>;
  debugGetSessionStatus(): Promise<JsonObject>;
  debugSetBreakpoint(location: JsonObject): Promise<JsonObject>;
  debugListBreakpoints(): Promise<JsonObject>;
  debugClearBreakpoints(): Promise<JsonObject>;
  debugContinue(timeoutS?: number): Promise<JsonObject>;
  debugHalt(timeoutS?: number): Promise<JsonObject>;
  debugGetStopReason(): Promise<JsonObject>;
  debugSymbolInfo(symbol: string): Promise<JsonObject>;
  debugDumpSymbolIhex(symbol: string, output: JsonObject): Promise<JsonObject>;
  classifyLastError(): Promise<JsonObject>;
  close(): Promise<void>;
}

export function createDebuggerBackend(config: AIHILConfig): DebuggerBackend {
  if (config.debugger.type === "openocd") {
    return new OpenOCDBackend(config);
  }
  if (config.debugger.type === "stlink") {
    return new STLinkBackend(config);
  }
  throw new ConfigError("config_invalid", "Unsupported debugger.type.", {
    field: "debugger.type",
    value: config.debugger.type,
    allowed_values: ["openocd", "stlink"],
  });
}
